user_modules/api/APIHelper.lua

```diff
@@ -0,0 +1,195 @@
+}
+
+--[[ Functions ]]-----------------------------------------------------------------------------------
+
+--[[ @getType: Provide type of the parameter without interface
+--! @parameters:
+--! pType: type of the parameter (possibly with interface)
+--! @return: type of the parameter
+--]]
+local function getType(pType)
+  if string.find(pType, "%.") then
+    return utils.splitString(pType, ".")[2]
+  end
+  return pType
+end
+
+--[[ @getParamsData: Provide hierarchy of parameters and sub-parameters for the defined API function
+--! @parameters:
+--! pAPIType: type of the API, e.g. 'mobile' or 'hmi'
+--! pEventType: type of the event, e.g. 'request', 'response' or 'notification'
+--! pFuncName: name of the API function, e.g. 'GetVehicleData'
+--! @return: structure with hierarchy of parameters
+--]]
+local function getParamsData(pAPIType, pEventType, pFuncName)
+
+  local function buildParams(pTbl, pParams)
+    for k, v in pairs(pParams) do
+      pTbl[k] = utils.cloneTable(v)
+      if schema[pAPIType].struct[getType(v.type)] then
+        pTbl[k].data = {}
+        buildParams(pTbl[k].data, schema[pAPIType].struct[getType(v.type)].param)
+        pTbl[k].type = m.dataType.STRUCT.type
+      elseif schema[pAPIType].enum[getType(v.type)] then
+        pTbl[k].data = {}
+        for kk in utils.spairs(schema[pAPIType].enum[getType(v.type)]) do
+          table.insert(pTbl[k].data, kk)
+        end
+        pTbl[k].type = m.dataType.ENUM.type
+      end
+    end
+  end
+
+  local function getAPIParams()
+    if pAPIType == m.apiType.MOBILE then
+      return schema.mobile.type[pEventType].functions[pFuncName].param
+    elseif pAPIType == m.apiType.HMI then
+      local iName = utils.splitString(pFuncName, ".")[1]
+      local fName = utils.splitString(pFuncName, ".")[2]
+      return api.hmi.interface[iName].type[pEventType].functions[fName].param
+    end
+  end
+
+  local params = getAPIParams(pAPIType)
+
+  local out = {}
+  buildParams(out, params)
+
+  local function updateBooleanValue(pValue)
+    local o = pValue
+    if type(pValue) == "string" then
+      if pValue == "true" then o = true
+      elseif pValue == "false" then o = false
+      end
+    end
+    return o
+  end
+
+  local function updateValues(pTbl)
+    for k, v in pairs(pTbl) do
+      if type(v) == "table" then
+        updateValues(v)
+      else
+        pTbl[k] = updateBooleanValue(v)
+      end
+    end
+  end
+  updateValues(out)
+
+  return out
+end
+
+--[[ @getGraph: Provide graph of all parameters for defined API function
+--! @parameters:
+--! pAPIType: type of the API, e.g. 'mobile' or 'hmi'
+--! pEventType: type of the event, e.g. 'request', 'response' or 'notification'
+--! pFuncName: name of the API function, e.g. 'GetVehicleData'
+--! @return: graph of all parameters (incl. sub-parameters) for the API function
+--]]
+function m.getGraph(pAPIType, pEventType, pFuncName)
+
+  local function getGraph(pParams, pGraph, pParentId)
```

Comment:
From what I can tell this is just flattening out the map of all the parameters into a single array. 

Is there any reason for this approach? It seems to just make all of the logic more complex and confusing as far as I can tell.

Reply:
The main benefit of this approach is to have possibility to check not only `root` parameters (as before), but also other sub-parameters on any level of hierarchy.
E.g. in case of root VD parameter `bodyInformation` it's possible to check all children sub-parameters:
```
- bodyInformation
  - roofStatuses
    - location
      - rowspan
```

Reply:
Seems to me that a recursive approach would be easier to write and less confusing. I don't really see any benefit to having all of these parameters on the same level.

Reply:
I apologize if I was not describe the benefits of the approach in a clear way. Let me add some details.
For example we have the following structure:
```
(1) - bodyInformation
(2)   - roofStatuses
(3)     - location
(4)       - rowspan
```

Old approach doesn't allow us to check any parameter except of the root one (`bodyInformation`). All children parameters (`roofStatuses`, `location`, `rowspan`) would be defined with some valid random values and that's it.

Within new approach each sub-parameter of hierarchy is verified against API restrictions.

E.g. we're able to check `rowspan` parameter which is on the lowest level of hierarchy against `maxvalue="100"` API restriction.

In order to do so we need to build the minimal version of hierarchy of parameters up to the root.
This hierarchy tree should include mandatory parameters on each levels and sub-levels in order to have a valid scenario.
That's why we need all these functions: `getParents`, `getMandatoryNeighbors`, `getMandatoryChildren`
All the parameters in the tree should have valid values defined.

Then we can test `rowspan` parameter against any other restriction defined in API.

New approach iterates through all the levels of hierarchy and builds all possible cases depends on rules defined.
For example it's possible to build the cases for Arrays only, for Strings, for Mandatory etc.

Reply:
But this isn't the only way to check subparameters, correct? You should still be able to do so by moving recursively through the parameter map, and I'd argue doing so would be simpler and more intuitive. Flattening out the map into an array seems like an unnecessary step.

Is there any part of the logic that wouldn't be possible just using a map of all the parameters?

Reply:
I would say flattened map (graph) is main concept of the new approach.
This artificial simplified representation of API structures is much easier to use along with iterative functions implemented in:
 - `APITestDataGenerator` module to build a tree of parameters and their values.
 - `VD common` module to generate test cases.

Taking into account complexity of the task (testing of all VD parameters and sub-parameters by different test types) it would take considerable time to re-work and re-test a new approach with recursion and without flattened map.

Reply:
After looking through the code again, I can see there are a few places where it would be fairly difficult to refactor (mainly in logic tied to the `getTestCases` function). 

It does seem like the code isn't quite as confusing once you fully understand the structure of the graph, but I think the issue is that the structure is not terribly obvious to anyone reading the code, so I think this approach would be more acceptable if the code were commented better. Specifically, I would suggest adding clarifying comments throughout the `getTestCases` function, as well as a more detailed explanation of the general structure of the graph in the description for the `getGraph` function (such as the fact that it includes all subparameters as well as root-level parameters).

Reply:
Please find update in afb7faf